imports/server/activityTracking.ts

```diff
@@ -0,0 +1,118 @@
+import { Meteor } from 'meteor/meteor';
+import { MongoInternals } from 'meteor/mongo';
+import { Random } from 'meteor/random';
+import { ACTIVITY_GRANULARITY } from '../lib/config/activityTracking';
+import PuzzleActivities from './models/ConsolidatedActivities';
+import RecentActivities from './models/RecentActivities';
+import { RecentActivityType } from './schemas/RecentActivity';
+
+interface ActivityKey {
+  ts: Date;
+  hunt: string;
+  puzzle: string;
+}
+
+function serializeActivityKey({ ts, hunt, puzzle }: ActivityKey) {
+  return `${ts.getTime()}-${hunt}-${puzzle}`;
+}
+
+function deserializeActivityKey(key: string) {
+  const [ts, hunt, puzzle] = key.split('-');
+  if (!ts || !hunt || !puzzle) {
+    throw new Error(`Invalid activity key: ${key}`);
+  }
+  return { ts: new Date(parseInt(ts, 10)), hunt, puzzle };
+}
+
+async function fetchAndDeleteRecentActivity(
+  cutoff: Date,
+  session: InstanceType<typeof MongoInternals.NpmModules.mongodb.module.ClientSession>,
+) {
+  if (!session) {
+    throw new Error('Must be called with a session');
+  }
+
+  const usersByKeyAndType = new Map<string, Map<RecentActivityType['type'], Set<string>>>();
+  while (true) {
+    /* eslint-disable no-await-in-loop */
+    const recent = await RecentActivities.rawCollection().findOneAndDelete(
+      { ts: { $lt: cutoff } } as any,
+      { session }
+    );
+    if (!recent.value) {
+      break;
+    }
+
+    const key = serializeActivityKey(recent.value);
+    const usersByType = usersByKeyAndType.get(key) ?? new Map();
+    const users = usersByType.get(recent.value.type) ?? new Set();
+    users.add(recent.value.user);
+    usersByType.set(recent.value.type, users);
+    usersByKeyAndType.set(key, usersByType);
+    /* eslint-enable no-await-in-loop */
+  }
+
+  return usersByKeyAndType;
+}
+
+async function consolidateRecentActivity() {
+  const ageForConsolidation = ACTIVITY_GRANULARITY * 2;
+  const cutoff = new Date(Date.now() - ageForConsolidation);
+
+  // By using a transaction, we avoid races with other servers, since we only
+  // count records that we delete (and if we fail to update the consolidated
+  // counters, the deletions will be rolled back).
+  //
+  // However: note that using the raw collection bypasses our schema
+  // enforcement, so we need to make sure of compliance
+  const { client } = MongoInternals.defaultRemoteCollectionDriver().mongo;
+  const session = client.startSession();
+  await session.withTransaction(async () => {
```

Comment:
What exactly happens here if the transaction aborts?  Does this backend throw a top-level exception that gets ignored, but then fails to reschedule?  Does it retry the transaction?  Something else?

Reply:
`withTransaction` is smart enough to retry the transaction if it knows that the transaction is retryable (I think this mostly happens in the event of a primary failover), but otherwise will rollback and throw. I think that would block rescheduling, so yes this should probably catch the exception so that we continue to re-schedule the consolidation.